Implemented SetupPremake.py to Download the Latest Premake Release

manoyes opened the pull request:
#### Describe the issue (if no issue has been made)

Closes #545

#### PR impact 
List of related issues/PRs this will solve:

 Impact                  | Issue/PR
------------------------ | ------
Issues this solves       |  Closes #545
Other PRs this solves    | None

#### Proposed fix 

Implemented SetupPremake.py to download ~`5.0.0-beta1`~ the latest version of premake instead of using `5.0.0-alpha16`. If the lookup fails, it will default to `5.0.0-beta1`.

In addition to being more stable, the beta adds support for `Visual Studio 2022` should people wish to use it. This allows me to natively build projects for `VS22` by manually editing `Win-GenProjects.bat`. I have not edited `Win-GenProjects.bat` to avoid breaking the build for other devs who are not using `VS22`, but downloading the beta rather than the alpha gives everyone the option to use `VS22` natively, and also generally adds stability over using the alpha.

#### Additional context

I tested this solution on Windows 10 Home 64-bit, it appears to download premake correctly and I was able to build and run the engine.

[Premake 5.0.0-beta1 Release and Changelog](https://github.com/premake/premake-core/releases/tag/v5.0.0-beta1)

nhathuy13598 commented:
https://github.com/TheCherno/Hazel/blob/b46de4f0f258aa1d0a7e79717c974c43fcc3f785/scripts/SetupPremake.py#L8
Instead of defining constant version, I think we should get the latest version of premake automatically. This is the `GetLatestPremakeVersion` function I made

```
import requests
...
@classmethod
def GetLatestPremakeVersion(self):
    LATEST_ENTRY = 0
    TAG_FIELD = "tag_name"
    url = "https://api.github.com/repos/premake/premake-core/releases"
    headers = {
        'Accept': 'application/vnd.github.v3+json'
    }
    response = requests.request("GET", url, headers = headers)
    if (response.status_code != 200) :
        return "5.0.0-beta1"
    return response.json()[LATEST_ENTRY][TAG_FIELD][1:] # Strip 'v' character out of tag name. Ex: v5.0.0-beta1 -> 5.0.0-beta1
```

manoyes commented:
@nhathuy13598 did you want to submit your own pull request or have me update mine?

nhathuy13598 commented:
> tag_name

You can update your pull request

manoyes commented:
@nhathuy13598 I updated the pull request with your function. See the commit msg for details:

```
Add Premake version auto lookup

Update the SetupPremake.py to download the latest available Premake
release by doing a GET request for release metadata in JSON format
and getting the first tag name. It strips out the leading v from the
semantic versioning. If the request fails it will default to the
5.0.0-beta1.

Credit for the GET request function goes to @nhathuy13598. I modified
their code by adding a comment and removing the classmethod directive,
as the GetLatestPremakeVersion function needed to be declared before
the class variables. I elected to define this function outside the
PremakeConfiguration class to keep the class variables at the top of
the file.

Closes https://github.com/TheCherno/Hazel#545
```
